Add verify subcommand to 'ansible-galaxy collection'

s-hertel opened the pull request:
##### SUMMARY
Adds the command `ansible-galaxy collection verify` to validate that the content of an installed collection is the same as the content of the collection on the server.

Note that this does not
  * verify dependencies listed in the requirements.yml
  * verify dir/file permissions
  * verify installations against tar.gz paths/urls

A requirements.yml file or multiple collections can be verified. When verifying multiple collections and one has a fatal error you can use --ignore-errors to continue verifying the rest.

The collection name argument for `ansible-galaxy collection verify` needs to be in the namespace.name format.

Some examples using the installed collection newswangerd.collection_demo:1.0.4

When verifying namespace.name without a version, it currently defaults to the latest
```
$ ansible-galaxy collection verify newswangerd.collection_demo
newswangerd.collection_demo has the version '1.0.4' but is being compared to '1.0.10'
```

To see details about the installed collection path and the server path use -vvv
Successful verification is quiet unless -vvv is used
```
$ ansible-galaxy collection verify newswangerd.collection_demo:1.0.4
$
```

After tampering with the installed collection.
```
$ ansible-galaxy collection verify newswangerd.collection_demo:1.0.4
Collection newswangerd.collection_demo contains modified content in the following files:
newswangerd.collection_demo
    MANIFEST.json
    FILES.json
    plugins/modules/real_facts.py
```
With extra verbosity the collection path and hash mismatches are shown:
```
$ ansible-galaxy collection verify newswangerd.collection_demo:1.0.4 -vvv
<...>
Collection newswangerd.collection_demo contains modified content in the following files:
newswangerd.collection_demo
/home/user/ansible/sample_collections/ansible_collections/newswangerd/collection_demo
    MANIFEST.json
    Expected: 6d1d1e56192673396e6cb04f92ab9e8cb3879d10e98bd8a6d9d5721d7b6d6ff1
    Found: 8ba47b3b6132cd1bf19063f474d50df3d5a19fc892055a298c5c203272781eea
    FILES.json
    Expected: 20f4ed17df2958eb48a6f3a32c99094b50566b95e8baaa31e17d31668607aa7d
    Found: 3405c19acc8d294e77677f06bdb7471717941b75074aef471d30d7ff1c9a83f8
    plugins/modules/real_facts.py
    Expected: 5f27fdc12026868f1f265f76a2aaecb8a0bcc353864ca36cd665f8af9392728e
    Found: 12186c36ce22f692a4e95444973c16e1bc494849330019ef33e81f599e743e88
```

Trying to verify a collection that doesn't exist on one of the enabled galaxy servers is fatal but can be made a warning with --ignore-errors:
```
$ ansible-galaxy collection verify test.doesnotexistonserver --ignore-errors
[WARNING]: Failed to verify collection test.doesnotexistonserver but skipping due to --ignore-errors being set. Error: Failed to find remote collection test.doesnotexistonserver:*
```

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
lib/ansible/cli/galaxy.py
lib/ansible/galaxy/collections.py

s-hertel committed:
[WIP] Add verify subcommand command to 'ansible-galaxy collection'

s-hertel committed:
Fix pep8 and use consistent display order

s-hertel committed:
WIP - docs

s-hertel committed:
Remove some redundancy in verify display messages by using an error queue for each collection

s-hertel committed:
Share common code and improve output format

s-hertel committed:
clean up documentation

s-hertel committed:
typo

samccann commented on docs/docsite/rst/user_guide/collections_using.rst:
```diff
@@ -47,6 +47,58 @@ Configuring the ``ansible-galaxy`` client
+
+   ansible-galaxy collection verify my_namespace.my_collection
+
+The output of the `ansible-galaxy collection verify` command is quiet if it is successful. If a collection has been modified, the altered files are listed under the collection name.
```
```suggestion
The output of the ``ansible-galaxy collection verify`` command is quiet if it is successful. If a collection has been modified, the altered files are listed under the collection name.
```

samccann commented on docs/docsite/rst/user_guide/collections_using.rst:
```diff
@@ -47,6 +47,58 @@ Configuring the ``ansible-galaxy`` client
+        plugins/inventory/my_inventory.py
+        plugins/modules/my_module.py
+
+You can use the -vvv flag to display additional verbosity, such as the version and path of the installed collection, the URL of the remote collection used for validation, and successful verification output.
```
```suggestion
You can use the ``-vvv`` flag to display additional information, such as the version and path of the installed collection, the URL of the remote collection used for validation, and successful verification output.
```

samccann commented on docs/docsite/rst/user_guide/collections_using.rst:
```diff
@@ -47,6 +47,58 @@ Configuring the ``ansible-galaxy`` client
+
+   ansible-galaxy collection verify my_namespace.my_collection:1.0.0
+
+In addition to the my_namespace.collection_name:version format, you can provide the collection to verify as a path or URL to a tar.gz file, a directory to an installed collection, or in a requirements.yml file. Dependencies listed in requirements.yml are not included in the verify process and should be verified separately.
```
```suggestion
In addition to the my_namespace.collection_name:version format, you can provide the collection to verify as a path or URL to a tar.gz file, a directory to an installed collection, or in a ``requirements.yml`` file. Dependencies listed in ``requirements.yml`` are not included in the verify process and should be verified separately.
```

s-hertel committed:
Apply suggestions from code review

Co-Authored-By: Sandra McCann <samccann@redhat.com>

sivel commented on lib/ansible/galaxy/collection.py:
```diff
@@ -186,6 +187,66 @@ def set_latest_version(self):
+            actual_hash = _consume_file(file_object)
+
+        if expected_hash != actual_hash:
+            ModifiedContent = namedtuple('ModifiedContent', ['filename', 'expected', 'installed'])
```
I'd define this in the "global" scope, so that it's defined just once and they are all the same.

sivel commented on lib/ansible/galaxy/collection.py:
```diff
@@ -461,6 +522,60 @@ def validate_collection_name(name):
+                    located_by_name = False
+
+                    if os.path.isfile(b_collection):
+                        remote_collection = CollectionRequirement.from_tar(to_bytes(collection[0]), False, parent=None)
```
You have access to `b_collection` here, just use that instead of re-converting `to_bytes`.

sivel commented on lib/ansible/galaxy/collection.py:
```diff
@@ -461,6 +522,60 @@ def validate_collection_name(name):
+                    if b_temp_tar_path is None:
+                        if not located_by_name:
+                            remote_collection = CollectionRequirement.from_name(collection_name, apis, collection_version, False, parent=None)
+                        remote_collection._get_metadata()
```
This may impact @samdoran as well, but if this is useful outside of the `CollectionRequirement` or not called automatically, then we might want to consider removing the leading `_`

sivel commented on lib/ansible/galaxy/collection.py:
```diff
@@ -461,6 +522,60 @@ def validate_collection_name(name):
+                        if not located_by_name:
+                            remote_collection = CollectionRequirement.from_name(collection_name, apis, collection_version, False, parent=None)
+                        remote_collection._get_metadata()
+                        download_url = remote_collection._metadata.download_url
```
This too, seems both `verify` and `list` need access to the metadata, so we might want to remove the leading `_`

s-hertel replied:
I added a public property `metadata` that calls `_get_metadata` if it hasn't been set yet. Do you prefer just removing the leading `_`s?

sivel replied:
No, that is fine.  Having a public property/method is all I was really after.

sivel commented on lib/ansible/galaxy/collection.py:
```diff
@@ -928,3 +1022,49 @@ def _extract_tar_file(tar, filename, b_dest, b_temp_path, expected_hash=None):
+    return _tarfile_extract(tar, member)
+
+
+def _get_tar_file_as_json(b_path, filename):
```
Not super necessary, but the name of this function is misleading to me.  Maybe `_get_json_from_tar_file` or something similar :)

s-hertel committed:
Move ModifiedContent namedtuple to the global scope

Add a public metadata property

Rename function to _get_json_from_tar_file

samccann commented:
This also needs a changelog fragment and consider a porting guide entry as well?

s-hertel committed:
Add some unit tests

s-hertel committed:
fix using common functions after rebase

s-hertel committed:
changelog

s-hertel committed:
Improve efficiency finding specific installed collections

Improve efficiency by only downloading the tar.gz from the galaxy server for comparison after checking that the collection has been installed

Handle multiple collection paths

Fix up tests

s-hertel committed:
pep8

s-hertel committed:
reword that for accuracy

samdoran commented on lib/ansible/cli/galaxy.py:
```diff
@@ -110,6 +111,7 @@ def init_parser(self):
         self.add_build_options(collection_parser, parents=[common, force])
         self.add_publish_options(collection_parser, parents=[common])
         self.add_install_options(collection_parser, parents=[common, force])
+        self.add_verify_options(collection_parser, parents=[common])
```
Could you use the `collections_path` parents so to save you from having to add the `-p` argument in `add_verify_options()`?
```suggestion
        self.add_verify_options(collection_parser, parents=[common, collections_paths])
```

samdoran commented on lib/ansible/galaxy/collection.py:
```diff
@@ -479,6 +541,62 @@ def validate_collection_path(collection_path):
+
+                    local_collection = None
+                    b_temp_tar_path = None
+                    b_collection = to_bytes(collection[0], errors='surrogate_or_strict')
```
The collection name needs to be verified, otherwise there will be a stack trace on line 565.
```
> ansible-galaxy collection verify bad -vvv
ansible-galaxy 2.10.0.dev0
  config file = /Users/sdoran/.ansible.cfg
  configured module search path = ['/usr/local/share/ansible/library']
  ansible python module location = /Users/sdoran/Source/ansible/lib/ansible
  executable location = /Users/sdoran/Source/ansible/bin/ansible-galaxy
  python version = 3.7.4 (default, Oct 15 2019, 10:30:33) [Clang 11.0.0 (clang-1100.0.33.8)]
Using /Users/sdoran/.ansible.cfg as config file
ERROR! Unexpected Exception, this is probably a bug: not enough values to unpack (expected 2, got 1)
the full traceback was:

Traceback (most recent call last):
  File "/Users/sdoran/Source/ansible/bin/ansible-galaxy", line 123, in <module>
    exit_code = cli.run()
  File "/Users/sdoran/Source/ansible/lib/ansible/cli/galaxy.py", line 398, in run
    context.CLIARGS['func']()
  File "/Users/sdoran/Source/ansible/lib/ansible/cli/galaxy.py", line 841, in execute_verify
    verify_collections(requirements, resolved_paths, self.api_servers, (not ignore_certs), ignore_errors)
  File "/Users/sdoran/Source/ansible/lib/ansible/galaxy/collection.py", line 565, in verify_collections
    remote_collection = CollectionRequirement.from_name(collection[0], apis, collection[1], False, parent=None)
  File "/Users/sdoran/Source/ansible/lib/ansible/galaxy/collection.py", line 371, in from_name
    namespace, name = collection.split('.', 1)
ValueError: not enough values to unpack (expected 2, got 1)
```

Something like this should do the trick.
```suggestion
                    verify_collection_name(collection[0])
                    b_collection = to_bytes(collection[0], errors='surrogate_or_strict')
```

samdoran reviewed (changes requested):
I found one minor bug when an invalid FQCN is provided. Docs and tests look great!

One thing I noticed is this doesn't seem to be able to verify _local only_ collections that don't exist on a remote server. This may be an intentional design. Did we plan to verify collections that _don't_ exist on a remote server even if they have a `FILES.json` with valid checksums? If not, we may need to adjust the error message a bit. I can see someone being confused by `ERROR! Failed to find collection samdoran.hotpotato:*` when the collection shows up in `ansible-galaxy collection list`.

s-hertel committed:
use more common code and verify collection name

s-hertel committed:
Improve error message and add documentation

s-hertel committed:
Update unit tests and add coverage for both local and remote collections that are unable to be located

AlanCoding commented:
Coming around just to test this. I created this repo:

https://github.com/AlanCoding/collection-dependencies-demo

This is a very basic collection. With it, I do the standard build then install commands. I ran this while the current version on Galaxy was 1.0.0.

```
$ ANSIBLE_COLLECTIONS_PATHS=~/Documents/repos/collection-dependencies-demo/target/ ansible-galaxy collection verify alancoding.deps
alancoding.deps has the version '1.0.1' but is being compared to '1.0.0'
```

So then I uploaded the new version to Galaxy.

```
$ ANSIBLE_COLLECTIONS_PATHS=~/Documents/repos/collection-dependencies-demo/target/ ansible-galaxy collection verify alancoding.deps
$ echo $?
0

```

Looks like it's working.

Trying the same with the .tar.gz file, from when the Galaxy version was still 1.0.0

```
$ ANSIBLE_COLLECTIONS_PATHS=~/Documents/repos/collection-dependencies-demo/target/ ansible-galaxy collection verify ~/Documents/repos/collection-dependencies-demo/alancoding-deps-1.0.1.tar.gz 
ERROR! Failed to find remote collection alancoding.deps:1.0.1 on any of the galaxy servers
```

Looks good. Here it is after I built the new version and uploaded it, but had not yet installed it.

```
$ ANSIBLE_COLLECTIONS_PATHS=~/Documents/repos/collection-dependencies-demo/target/ ansible-galaxy collection verify ~/Documents/repos/collection-dependencies-demo/alancoding-deps-1.0.1.tar.gz 
ERROR! Unexpected Exception, this is probably a bug: [Errno 2] No such file or directory: b'/Users/alancoding/Documents/repos/collection-dependencies-demo/target/ansible_collections/alancoding/deps/README.md'
```

hrmm... after installing it:

```
$ ANSIBLE_COLLECTIONS_PATHS=~/Documents/repos/collection-dependencies-demo/target/ ansible-galaxy collection verify ~/Documents/repos/collection-dependencies-demo/alancoding-deps-1.0.1.tar.gz 
$ echo $?
0

```

----

Everything is good except for the 1 case where it failed with the `alancoding-deps-1.0.1.tar.gz` on the `README.md` file.

It's verifying the .tar.gz file, but it gives an error about the installed version. Those are two very different things. Is it verifying the install, or verifying the build? It looks like it's verifying both, and this seems like a problem for people integrating this into release pipelines and stuff like that.

s-hertel committed:
Only validate collections using the format collection_namespace.collection_name

Fix tests to reflect that

Fix documentation

s-hertel committed:
Fix a bug when a local collection does not contain a file

s-hertel commented:
@AlanCoding Thanks for testing, your feedback was really helpful. I've fixed the bug you found and eliminated verifying anything but collections in the format namespace.name to make it very clear that it's intended to verify installations, not builds.

s-hertel committed:
pep8

AlanCoding commented:
confirmed, this is more clear behavior to me

```
$ ANSIBLE_COLLECTIONS_PATHS=~/Documents/repos/collection-dependencies-demo/target/ ansible-galaxy collection verify ~/Documents/repos/collection-dependencies-demo/alancoding-deps-1.0.1.tar.gz
ERROR! '/Users/alancoding/Documents/repos/collection-dependencies-demo/alancoding-deps-1.0.1.tar.gz' is not a valid collection name. The format namespace.name is expected.
```

In retrospect, I realize that something conceptually whooshed over my head. It would be a coherent idea to validate an installed collection against a `*.tar.gz` file (or a local cache of multiple of them). But then you have the difficulty of communication of validating the install versus the build, or the build versus the server. It's not claiming to do either of those, and that makes sense to me. 👍  from me

I did want to share 1 more output to show the specific usefulness:

```
$ ansible-galaxy collection verify awx.awx:9.1.0
Collection awx.awx contains modified content in the following files:
awx.awx
    MANIFEST.json
    FILES.json
    galaxy.yml.j2
    plugins/modules/tower_job_template.py
    plugins/modules/tower_project.py
    plugins/modules/tower_role.py
    setup.cfg
    template_galaxy.yml
```

I was doing development for the `awx.awx` collection, and (apparently) half-way through its 9.1.0 release, I had installed it locally. This verify command accurately showed exactly which files had modifications in the final released version compared to what I had on my computer at that time. Kind of cool.

samdoran commented on docs/docsite/rst/user_guide/collections_using.rst:
```diff
@@ -47,6 +47,54 @@ Configuring the ``ansible-galaxy`` client
+
+   ansible-galaxy collection verify -r requirements.yml
+
+Verifying against tar.gz files is not supported. If your ``requirements.yml`` contains paths to tarfiles or URLs for installation, you can use the ``--ignore-errors`` flag to ensure that all collections using the namespace.name format in the file are processed.
```
```suggestion
Verifying against ``tar.gz`` files is not supported. If your ``requirements.yml`` contains paths to tar files or URLs for installation, you can use the ``--ignore-errors`` flag to ensure that all collections using the ``namespace.name`` format in the file are processed.
```

s-hertel replied:
Thanks, good spotting. Fixed one I missed above too.

s-hertel committed:
Fix formatting

s-hertel commented:
bot_status

sivel commented:
Happy merge day!
